dnprogs/apps/dnping.1
.TH DNPING 1 "January 25 2000" "DECnet utilities"

.SH NAME
dnping \- Loopbacks diagnostic packets through a remote node

.SH SYNOPSIS
.B dnping
nodename [user pass] count
.br
or
.br
.B dnping
nodename [options] nodename
.br
Options:
.br
[\dqsv] [\-c number] [\-i interval] [\-p password] [\-s size] [\-u username] [\-w timeout]
.br
.SH DESCRIPTION
.PP
This utility sends to remote DECnet node
.B nodename
the number of packets specified by
.B count
to test the link between the two systems. Optionally a
username and password may be specified for the connection as well as
several other options. NOTE that if you dnping another Linux box it must
have
.B dnetd
running.
.br
NOTE also that dnping is not really like an IP "ping" in that it needs
a registered object at the other end to connect to. So, just because
you cannot ping a machine does not, necessarily, mean that machine
is not available, just that the MIRROR object is not available. There
is not (to my knowledge) a low-level equivalent in DECnet of the ICMP
ping message.

.SH OPTIONS
.TP
.I "\-c number"
Number of packets to send (default 10)
.TP
.I "\-d"
Debug mode (default off)
.TP
.I "\-i interval"
interval between packets in microseconds (default 0)
.TP
.I "\-p password"
Access control password. If this is "-" then you will be prompted.
.TP
.I "\-q"
Quiet mode (default off)
.TP
.I "\-s size"
size of frame to send in bytes (40 data + 68 hdr)
.TP
.I "-t"
timestamps mode (default off)
.TP
.I "-u username"
access control username
.TP
.I "-w timeout"
Specifies a timeout (in seconds). If not response is received
after this time then dnping will abort. The default is to wait
forever.
.TP
.I "-v"
verbose mode (default off)

.SH EXAMPLES
.br
Pings 10 packets through remote node "mv3100"
.br

.br
.PP
    # dnping mv3100 10
.br
.br
Make it look a bit like IP ping:
.PP
    # dnping  \-vti 1000000 marsha
.br

.SH SEE ALSO
.BR dntype "(1), " dndir "(1), " dndel "(1), " dntask "(1), "
.BR sethost "(1), "  dnetd "(8)"


dnprogs/NEWS
dnprogs (2.61) unstable; urgency=low

  * Include libvaxdata/src/test.c in the distribution tarball
  * Fix password length in DAP connections. It was incorrectly set at 12 
    rather than 40
 
 -- Chrissie Caulfield <chrissie@debian.org>  Sun, 31 Mar 2013 11:18:32 +0000

dnprogs (2.60+x32) unreleased; urgency=low

  * Switch from using deprecated sysctl(2) to writing to /proc/sys.  This
    fixes a build failure on x32.

 -- Daniel Schepler <schepler@debian.org>  Tue, 12 Feb 2013 12:21:31 -0800

dnprogs (2.60) unstable; urgency=low

  * Fixed lots of Debian things that caused lintian problems and 
    upgraded policy version
    Closes: #673408
  * Fix some warnings that might be bugs
  * Incorporate NMU into main tree.

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 26 May 2012 09:54:41 +0100

dnprogs (2.59+nmu2) unstable; urgency=medium

  * NMU
  * dnprogs binaries are Linux-only, so state that. Move dnet-common back
    to binary-all. Closes: #665886

 -- Steve McIntyre <93sam@debian.org>  Mon, 21 May 2012 17:43:14 +0100

dnprogs (2.59+nmu1) unstable; urgency=low

  * NMU
  * Reduce libdnet's recommends on dnet-common to suggests again.
    Closes: #655740. Adapted patch from Jonathan Nieder.

 -- Steve McIntyre <93sam@debian.org>  Fri, 18 May 2012 13:29:53 +0100

dnprogs (2.59) unstable; urgency=low

  * Changed package dnet-common from arch all to any and
    updated dependecies. (Closes: #665886)

 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 27 Apr 2012 09:30:19 +0100

dnprogs (2.58) unstable; urgency=medium

  * Fixed lintian warnings:
    - Fixed changelog entry of last release to have correct line length.
    - Corrected debconf dependency in dnet-common.
    - Corrected versioned dependency in dnprogs on dnet-common.
    - Updated existing overrides to match new syntax.
  * Updated linkage of libdnet_daemon and libdap. (Closes: #646353)
  * Fixed compile warning in dndir when using -Werror=format-security
    (Closes: #646613)
  * Fixed typo in debian/dnet-common.config
  * Do not try to install files into packages on FreeBSD which are
    not build on FreeBSD.
  * Link dapfs with libpthread for new libfuse compatibility
  * Updated nl.po (Closes: #652349)
  * Added pl.po (Closes: #661894)
  * Added it.po (Closes: #663135)
  * Change module-init-tools dependency to kmod (Closes: #662684)

 -- Chrissie Caulfield <chrissie@debian.org>  Sat, 10 Mar 2012 10:32:08 +0000

dnprogs (2.57) unstable; urgency=medium

  [ Chrissie Caulfield ]
  * Completed removal of modutils configuration file code
  * Fix presence of CVS directories in tarball
  * updated 'clean' target

  [ Philipp Schafft ]
  * Use /dev/pts PTYs on 64 bit systems (fix makefile bug)
  * Updated translations Debconf templates:
    - German translation (Closes: #615268)
    - French translation (Closes: #615039)
    - Danish translation (Closes: #633984) (added)
    - Slovak translation (Closes: #639629) (added)
  * Look in yet another place for libcrypt (better fix than from NMUs)
  * Added a comment about the detecting of libcrypt. (See: #629664)
  * "ethernet" was capitalized in extended description (Closes: #636798)
  * Fixed some lintian warnings (mainly new rules targets).
  * dnet-common: Do not configure DECnet unless the user requested it.
    (Removed the bad workaround done by NMU) (See: #637179)
  * dnet-common: Added preinst script to clean up the broken config files
    from the last NMU. (Closes: #640861)
  * Upgraded libdnet's suggest to dnet-common to recommend again
    (Closes: #608807)

 -- Chrissie Caulfield <chrissie@debian.org>  Thu,  8 Sep 2011 16:05:47 +0200

dnprogs (2.56.1+nmu1) unstable; urgency=low

  * Non-maintainer upload.
  * libdnet: Only suggest dnet-common package (Closes: #608807)
  * dnet-common: Do not change MAC addresses unless user chooses to
    configure DECnet (Closes: #637179)
  * dnet-common: Restore ability to configure DECnet name and address
    through debconf
  * dnet-common: Remove obsolete modutils configuration file

 -- Ben Hutchings <ben@decadent.org.uk>  Thu, 25 Aug 2011 14:26:12 +0100

dnprogs (2.56.1) unstable; urgency=low

  * Non-maintainer upload.
  * Add patch from Ubuntu to fix FTBFS with libcrypt moved to the
    multiarch path.

 -- Aurelien Jarno <aurel32@debian.org>  Tue, 09 Aug 2011 11:58:06 +0200


dnprogs (2.56) unstable; urgency=low

  * Search for libgcrypt in 64bit library paths, too. (thanks to Kamil)
  * Package dnet-common now asks if network interfaces should be configrued
    for DECnet (Closes: #608807)
    Updated translations Debconf templates:
    - Russian translation (Closes: #610642)
    - Portuguese translation (Closes: #611034)
    - Swedish translation (Closes: #611296)
    - Czech translation (Closes: #611430)
    - Japanese translation (Closes: #611448)
    - Spanish translation (Closes: #611547)
  * Build against version 25 of FUSE
    Closes: #598831
  * Fixed some lintian warnings (typos and such).
  * Raised priority of package libdnet from extra to optional
    to comply policy manual section 2.5 "Priorities".

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 22 Feb 2011 10:38:23 +0000

dnprogs (2.55) unstable; urgency=low

  * Build against version 25 of FUSE
    Closes: #598831

 -- Chrissie Caulfield <chrissie@debian.org>  Tue, 5 Oct 2010 09:23:12 +0100

dnprogs (2.54) unstable; urgency=low

  * Don't build libvaxdata test program by default.
    Closes: #598752
 
 -- Chrissie Caulfield <chrissie@debian.org>  Fri, 1 Oct 2010 20:44:09 +0100

dnprogs (2.53) unstable; urgency=low

  * Done some more porting to FreeBSD, corrected Build-Depends for hurd-i386
    Closes: #588185
  * Added new tool node(1) - Node name lookup tool
  * Added support to libdnet_daemon to read /etc/nodes.{allow,deny}
  * Fixed DoS bug in dnet_daemon()
    Closes: #592015 
  * Update libvaxdata to version 1.2, fixes double conversions

 -- Christine Caulfield <Christine.Caulfield@googlemail.com>  Thu, 19 Jul 2010 10:19:32 +0100


dnprogs/apps/dnping.c
/******************************************************************************
    (c) 1998 Eduardo.M Serrat             emserrat@hotmail.com
    Username/Password additions by David G North 1999
    Optarg additions by Rob Davies - Feb 2000

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.
******************************************************************************/
/* dnping.c */

#include <sys/types.h>
#include <stdio.h>
#include <string.h>
#include <limits.h>
#include <errno.h>
#include <unistd.h>
#include <stdlib.h>
#include <ctype.h>
#include <signal.h>
#include <sys/time.h>
#include <sys/socket.h>
#include <netdnet/dn.h>
#include <netdnet/dnetdb.h>

#define DNF_PKT_COUNT     0x001
#define DNF_INTERVAL      0x002
#define DNF_PKT_SIZE      0x004
#define DNF_USERNAME      0x008
#define DNF_PASSWORD      0x010
#define DNF_VERBOSE       0x020
#define DNF_QUIET         0x040
#define DNF_DEBUG         0x080
#define DNF_TIMESTAMPS    0x100
#define DNF_TIMEOUT       0x200

#define MIN(a,b) (((a) < (b)) ? (a) : (b))

#define MAX_DN_PACKETSIZE 1518     /* set Ethernet max as maximum size */
#define MAX_DN_HDRSIZE    68       /* this is a wild guess -
                                      it needs to be calculated properly */
#define MAX_DN_DATASIZE   (MAX_DN_PACKETSIZE - MAX_DN_HDRSIZE)

int   options = 0;

/*-------------------------------------------------------------------------*/
static void usage(void)
{
    printf("Usage:\n");
    printf("\ndnping nodename [user pass] count\n");
    printf("\n\t*or*\n");
    printf("\ndnping [options] nodename\n");
    printf("\twhere [options]:\n");
    printf("\t-c number      number of packets to send {10}\n");
    printf("\t-d             debug mode {OFF}\n");
    printf("\t-i interval    interval between packets in microseconds {0}\n");
    printf("\t-p password    access control password {}\n");
    printf("\t-q             quiet mode {OFF}\n");
    printf("\t-s size        size of frame to send in bytes {%d data + %d hdr}\n",
	   40,MAX_DN_HDRSIZE);
    printf("\t-t             timestamps mode {OFF}\n");
    printf("\t-u username    access control username {}\n");
    printf("\t-v             verbose mode {OFF}\n");
    printf("\t-w seconds     maximum wait time (timeout)\n");
    exit(0);
}

void tvsub(register struct timeval *out, register struct timeval *in)
{
    if ((out->tv_usec -= in->tv_usec) < 0)
    {
	--out->tv_sec;
	out->tv_usec += 1000000;
    }
    out->tv_sec -= in->tv_sec;
}


void init_accdata( char *user, char *password,
		   struct accessdata_dn *accessdata)
{
    char *local_user = cuserid(NULL);
    char *cp;

    if ((options & DNF_DEBUG) && (local_user))
    {
	printf("cuserid() - LOCAL USER: %s\n",local_user);
    }

    memset(accessdata, 0, sizeof(*accessdata));

    memcpy(accessdata->acc_user, user, MIN(strlen(user),DN_MAXACCL));
    accessdata->acc_user[DN_MAXACCL-1] = '\0';
    accessdata->acc_userl = strlen((char *)accessdata->acc_user);


    /* If the password is "-" and fd 0 is a tty then
       prompt for a password */
    if (password[0] == '-' && password[1] == '\0' && isatty(0))
    {
	password = getpass("Password: ");
	if (password == NULL || strlen(password) > (unsigned int)DN_MAXACCL)
	{
	    fprintf(stderr, "Password too long");
	    return;
	}

    }

    memcpy(accessdata->acc_pass, password, MIN(strlen(password),DN_MAXACCL));
    accessdata->acc_pass[DN_MAXACCL-1] = '\0';
    accessdata->acc_passl = strlen((char *)accessdata->acc_pass);

    /* Try very hard to get the local username for proxy access */
    if (!local_user || local_user == (char *)0xffffffff)
    {
	local_user = getenv("LOGNAME");

	if ((options & DNF_DEBUG) && (local_user))
	{
	    printf("getenv(LOGNAME) - LOCAL USER: %s\n",local_user);
	}

    }

    if (!local_user)
    {
	local_user = getenv("USER");

	if ((options & DNF_DEBUG) && (local_user))
	{
	    printf("getenv(USER) - LOCAL USER: %s\n",local_user);
	}

    }

    if (local_user)
    {
	strncpy((char *)accessdata->acc_acc, local_user,
		MIN(strlen(local_user),DN_MAXACCL));
	accessdata->acc_acc[DN_MAXACCL-1] = '\0';
	accessdata->acc_accl = strlen((char *)accessdata->acc_acc);
	for (cp=(char *)accessdata->acc_acc; *cp!='\0'; *cp=toupper(*cp), ++cp);
    }
    else
    {
	accessdata->acc_acc[0] = '\0';

	if (options & DNF_DEBUG)
	{
	    printf("LOCAL USER: NULL\n");
	}

    }

    return;
}

void sig_alarm(int s)
{
    printf("Connect timed out\n");
    exit(1);
}

/*-------------------------------------------------------------------------*/
int main(int argc, char *argv[])
{
    struct  sockaddr_dn             sockaddr;
    struct  accessdata_dn           accessdata;
    static  struct  nodeent         *np;
    int                     sockfd,i,ch;
    char                    nodename[20],
	ibuf[MAX_DN_PACKETSIZE],
	obuf[MAX_DN_PACKETSIZE];
    short                   snd,rcv,num;

    char                    username[DN_MAXACCL],password[DN_MAXACCL];
    int                     npackets = 10,
	datalen = 40,
	interval = 0,
	cmplen, offset;
    struct timeval          tv, *tp;
    long                    triptime = 0,
	tmin = LONG_MAX, /* minimum round trip time */
	tmax = 0;        /* maximum round trip time */
    unsigned long           tsum = 0; /* sum of all times, for doing average */
    unsigned long           timeout_sec;
    struct timeval          timeout;

    signal(SIGALRM, sig_alarm);

    while ((ch = getopt(argc, argv, "c:di:qs:u:p:w:vt")) != EOF)
    {
	switch(ch)
	{
	case 'c':               /* number of packets to send */
            npackets = atoi(optarg);
            if (npackets <= 0)
	    {
		fprintf(stderr, "ping: bad number of packets to transmit.\n");
		exit(-1);
	    }
            options |= DNF_PKT_COUNT;
            break;
	case 'd':               /* turn on debug option */
            options |= DNF_DEBUG;
            break;
	case 't':               /* turn on timestamps option */
            options |= DNF_TIMESTAMPS;
            break;
	case 'i':               /* wait between sending packets */
            interval = atoi(optarg);
            if ( (interval <= 0) || (interval > 60000000) )
	    {
		fprintf(stderr, "ping: bad timing interval.\n");
		exit(-1);
	    }
            options |= DNF_INTERVAL;
            break;
	case 'q':               /* quiet mode */
            options |= DNF_QUIET;
            break;
	case 's':               /* size of data portion to send */
            options |= DNF_PKT_SIZE;
            datalen = atoi(optarg) - MAX_DN_HDRSIZE;
            if (datalen > MAX_DN_DATASIZE)
	    {
		fprintf(stderr, "ping: packet size too large.\n");
		exit(-1);
	    }
            if (datalen <= 0)
	    {
		fprintf(stderr, "ping: illegal packet size.\n");
		exit(-1);
	    }
            break;
	case 'v':               /* verbose mode */
            options |= DNF_VERBOSE;
            break;
	case 'u':               /* access control username */
            options |= DNF_USERNAME;
            snprintf(username,sizeof(username),"%s",optarg);
            break;
	case 'p':               /* access control password */
            options |= DNF_PASSWORD;
            snprintf(password,sizeof(password),"%s",optarg);
            break;
	case 'w':
	    options |= DNF_TIMEOUT;
	    timeout_sec = atoi(optarg);
	    break;
	default:
            usage();
	    break;
	}
    }
    argc -= optind;
    argv += optind;

    if (options & DNF_DEBUG) printf("ARGC : %d\n",argc);

    if ((argc < 1) || (argc > 4))
    {
	usage();
    }

    snprintf(nodename,sizeof(nodename),"%s",*argv);
    if ( ( (argc == 4) || (argc == 2) )
	 && ((options & DNF_PKT_COUNT) == 0) )
    {
	npackets=atoi(argv[argc-1]);
    }

    if ( (np=getnodebyname(nodename)) == NULL)
    {
	if ( (options & DNF_QUIET) == 0 )
	{
	    printf("Unknown node name %s\n",nodename);
	}
	exit(-1);
    }

    if ((sockfd=socket(AF_DECnet,SOCK_SEQPACKET,DNPROTO_NSP)) == -1)
    {
	if ( (options & DNF_QUIET) == 0 )
	{
	    perror("socket");
	}
	exit(-1);
    }

    if ((((options & DNF_USERNAME) == 0) && ((options & DNF_PASSWORD) != 0)) ||
	(((options & DNF_USERNAME) != 0) && ((options & DNF_PASSWORD) == 0)))
    {
	if ( (options & DNF_QUIET) == 0 )
	{
	    printf("Must specify both username and password for access control\n");
	}
	exit(-1);
    }

    if ( ((options & (DNF_USERNAME|DNF_PASSWORD) ) == 0) && (argc > 2) )
    {
	snprintf(username,sizeof(username),"%s",argv[1]);
	options |= DNF_USERNAME;
	snprintf(password,sizeof(password),"%s",argv[2]);
	options |= DNF_PASSWORD;
    }

    if ( (options&(DNF_USERNAME|DNF_PASSWORD)) == (DNF_USERNAME|DNF_PASSWORD))
    {
	if ( (options & DNF_DEBUG) != 0)
	{
	    printf("USERNAME: %s\nPASSWORD: %s\n", username, password);
	}

	init_accdata(username, password, &accessdata);
	if (setsockopt(sockfd, DNPROTO_NSP, SO_CONACCESS, &accessdata,
		       sizeof(accessdata)) < 0)
	{
	    if ( (options & DNF_QUIET) == 0 )
            {
		perror("setsockopt");
            }
	    exit(-1);
	}
    }

    if (options & DNF_TIMESTAMPS)
    {
	if (datalen < sizeof(struct timeval))
	{
	    if ( (options & DNF_QUIET) == 0 )
            {
		printf("Packet size not large enough to store timestamp\n");
            }
	    exit(-1);
	}
    }

    sockaddr.sdn_family = AF_DECnet;
    sockaddr.sdn_flags  = 0x00;
    sockaddr.sdn_objnum  = DNOBJECT_MIRROR;
    sockaddr.sdn_objnamel  = 0x00;
    memcpy(sockaddr.sdn_add.a_addr, np->n_addr,2);

    /* This is the cheesy, cowards way of checking for
       a connect timeout */
    if (options & DNF_TIMEOUT)
	alarm(timeout_sec);

    if (connect(sockfd, (struct sockaddr *)&sockaddr,
		sizeof(sockaddr)) < 0)
    {
	if ( (options & DNF_QUIET) == 0 )
	{
	    perror("socket");
	}
	exit(-1);
    }

    /* Cancel the connect() alarm */
    if (options & DNF_TIMEOUT)
	alarm(0);


    for (i = 0; i < datalen; i++)
    {
	obuf[i]=0x85;
    }
    obuf[0]=0x00;

    cmplen = (datalen - 1) -
	(options & DNF_TIMESTAMPS)?sizeof(struct timeval):0;
    if (options & DNF_DEBUG)
    {
	printf("CMPLEN: %d\n",cmplen);
    }

    offset = 1 + (options & DNF_TIMESTAMPS)?sizeof(struct timeval):0;
    if (options & DNF_DEBUG)
    {
	printf("OFFSET: %d\n",offset);
    }

    snd=0;
    rcv=0;

    for (i = 0; i < npackets; i++)
    {
	if (options & DNF_TIMESTAMPS)
	{
	    gettimeofday((struct timeval *)&obuf[1], (struct timezone *)NULL);
	}
	num = write(sockfd,obuf,datalen);
	if ( num < 0 )
	{
	    if ( (options & DNF_QUIET) == 0 )
            {
		perror("Write");
            }
	    exit(-1);
	}
	if (options & (DNF_DEBUG|DNF_VERBOSE))
	{
	    printf("PKT: %-4d   WRITE: %d ",i+1,num);
	}
	snd++;

	if (options & DNF_TIMEOUT)
	{
	    int status;
	    fd_set in_fd;

	    FD_ZERO(&in_fd);
	    FD_SET(sockfd, &in_fd);
	    timeout.tv_sec = timeout_sec;
	    timeout.tv_usec = 0;

	    status = select(sockfd+1, &in_fd, NULL, NULL, &timeout);
	    if (status < 0)
	    {
		perror("select");
		close(sockfd);
		exit(-1);
	    }
	    if (status == 0)
	    {
		fprintf(stderr, "Timeout\n");
		close(sockfd);
		exit(-1);
	    }
	}

	num = read(sockfd,ibuf,sizeof(ibuf));
	if ( num < 0 )
	{
	    if ( (options & DNF_QUIET) == 0 )
            {
		perror("Read");
            }
	    exit(-1);
	}

	gettimeofday(&tv, (struct timezone *)NULL);

	if (options & (DNF_DEBUG|DNF_VERBOSE))
	{
	    printf("READ: %d ",num);
	}

	if (memcmp(&obuf[offset],&ibuf[offset],cmplen) != 0)
	{
	    if ( (options & (DNF_QUIET|DNF_DEBUG|DNF_VERBOSE)) == 0 )
            {
		printf("Loopback Error\n");
            }
	    if ( (options & (DNF_DEBUG|DNF_VERBOSE)) != 0)
            {
		printf("**error**\n");
            }
	}
	else
	{
	    rcv++;
	    if (options & DNF_TIMESTAMPS)
            {
		tp = (struct timeval *)&ibuf[1];
		tvsub(&tv, tp);
		triptime = tv.tv_sec * 10000 + (tv.tv_usec / 100);
		tsum += triptime;
		if (triptime < tmin)
		{
		    tmin = triptime;
		}
		if (triptime > tmax)
		{
		    tmax = triptime;
		}
            }

	    if ( (options & (DNF_DEBUG|DNF_VERBOSE)) != 0)
            {
		if (options & DNF_TIMESTAMPS)
		{
		    printf("  RTT: %ld.%ld\n", triptime/10, triptime%10);
		}
		else
		{
		    printf("\n");
		}
            }
	}

	if (interval > 0)
	{
	    usleep(interval);
	}

    } /* end for loop */

    close(sockfd);
    if ( (options & DNF_QUIET) == 0 )
    {
	printf("Sent %d packets, Received %d packets\n",snd,rcv);
	if ((rcv > 0) && (options & DNF_TIMESTAMPS))
	{
	    if (options & DNF_DEBUG)
            {
		printf("TSUM: %ld\n",tsum);
            }

	    printf("\tround-trip min/avg/max = %ld.%ld/%lu.%ld/%ld.%ld ms\n",
		   tmin/10, tmin%10,
		   (tsum / (rcv*10)),
		   (tsum / rcv)%10,
		   tmax/10, tmax%10);
	}
    }

    return( ((snd - rcv) != 0) ? -1 : 0 );
}
